src/client_side_request.cc

```diff
@@ -1863,6 +1889,106 @@ ClientHttpRequest::doCallouts()
 #endif
 }
 
+/// \return a new URI that honors uri_whitespace
+static char *
+cleanupUri(const char *uri)
+{
+    int flags = 0;
+    char *cleanedUri = nullptr;
+    switch (Config.uri_whitespace) {
+    case URI_WHITESPACE_ALLOW:
+        flags |= RFC1738_ESCAPE_NOSPACE;
+        // fall through to next case
+    case URI_WHITESPACE_ENCODE:
+        flags |= RFC1738_ESCAPE_UNESCAPED;
+        cleanedUri = xstrndup(rfc1738_do_escape(uri, flags), MAX_URL);
+        break;
+
+    case URI_WHITESPACE_CHOP: {
+        flags |= RFC1738_ESCAPE_UNESCAPED;
+        const auto pos = strcspn(uri, w_space);
+        char *choppedUri = nullptr;
+        if (pos < strlen(uri))
+            choppedUri = xstrndup(uri, pos + 1);
+        cleanedUri = xstrndup(rfc1738_do_escape(choppedUri ? choppedUri : uri, flags), MAX_URL);
+        cleanedUri[pos] = '\0';
+        xfree(choppedUri);
+    }
+    break;
+
+    case URI_WHITESPACE_DENY:
+    case URI_WHITESPACE_STRIP:
+    default: {
+        // TODO: avoid duplication with urlParse()
+        const char *t;
+        char *tmp_uri = static_cast<char*>(xmalloc(strlen(uri) + 1));
+        char *q = tmp_uri;
+        t = uri;
+        while (*t) {
+            if (!xisspace(*t)) {
+                *q = *t;
+                ++q;
+            }
+            ++t;
+        }
+        *q = '\0';
+        cleanedUri = xstrndup(rfc1738_escape_unescaped(tmp_uri), MAX_URL);
+        xfree(tmp_uri);
+    }
+    break;
+    }
+
+    assert(cleanedUri);
+    return cleanedUri;
+}
+
+void
+ClientHttpRequest::setLogUriToRequestUri()
+{
+    assert(request);
+    const auto canonicalUri = urlCanonicalClean(*request);
+    absorbLogUri(xstrndup(canonicalUri, MAX_URL));
+}
+
+void
+ClientHttpRequest::setLogUriToRawUri(const char *rawUri, const HttpRequestMethod &method)
+{
+    assert(rawUri);
+    // Should(!request);
```

Comment:
@yadij, this PR adds a few commented out Should() calls. The calls are useful when you want to check/document an assumption that is not necessary for the related code to "work" (so this is not an assertion or a Must()) but which violations are bugs that should be fixed. My idea was to provide (in a dedicated PR) a Should(condition) macro that will print a `BUG: ...` message when the condition is violated (once). We use a similar approach in Polygraph and other Factory code and like it. Do you think this is a good idea for Squid?

Reply:
I'd like to only claim a "BUG:" loudly if we are pretty certain that it is bad problem to be in that state. In which case its probably a Must() situation instead.

Making Should() only log at a low debugs() level would be okay though.

Reply:
> I'd like to only claim a "BUG:" loudly if we are pretty certain that it is bad problem to be in that state.

Yes, that is exactly what Should() is for.

> In which case its probably a Must() situation instead.

Not necessarily: It is often possible to be in a bad state and _not_ want to kill the current transaction, processing sequence, worker, etc. (which is what Must() and assert() do). Should() is not a Must().


@yadij, if you veto the Should() concept, then commented-out Should() invocations should be deleted from this PR. If the above explanation changes your mind about Should()s, then those commented-out invocations should stay (and I will submit a PR implementing Should()s). What say you?